File: Search/BinarySearch/TypeScript/binarySearch.ts
// The given array must be sorted!
const iterativeBinarySearch = (
  sortedList: number[],
  target: number,
): number | Error => {
  let low = 0
  let high = sortedList.length - 1

  while (low <= high) {
    const mid = Math.floor((low + high) / 2)
    // console.log(`Low: ${low}, Mid: ${mid}, High: ${high}`);

    if (sortedList[mid] === target) {
      // Return the value if the value in the middle is the same as the target value.
      return mid
    } else if (sortedList[mid] < target) {
      // Set low as (mid + 1) if the value in the middle is smaller that the target value.
      low = mid + 1
    } else {
      // Set high as mid if the value in the middle is bigger than the target value.
      high = mid - 1
    }
  }

  throw 'No such value'
}

const recursiveBinarySearch = (
  sortedList: number[],
  target: number,
  low: number,
  high: number,
): number | Error => {
  const mid = Math.floor((low + high) / 2)

  if (sortedList[mid] === target) return mid
  if (high < low) throw 'No such value'

  if (sortedList[mid] < target) {
    return recursiveBinarySearch(sortedList, target, mid + 1, high)
  } else {
    return recursiveBinarySearch(sortedList, target, low, mid - 1)
  }
}

export { iterativeBinarySearch, recursiveBinarySearch }

/* 
console.log(iterativeBinarySearch([1, 2, 3, 4, 5, 6, 7, 8, 9], 6));
Low: 0, Mid: 4, High: 8
Low: 5, Mid: 6, High: 8
Low: 5, Mid: 5, High: 6
*/

File: Search/BinarySearch/Ruby/binary_search.rb
class NoSuchValueError < StandardError
  def initialize(msg="No such value in the array")
    super
  end
end

def iterative_binary_search(sorted_list, target)
  low, high = 0, sorted_list.length

  while low < high
    mid = ((low + high) / 2).floor

    if sorted_list[mid] == target 
      return mid
    elsif sorted_list[mid] < target
      low = mid + 1
    else
      high = mid
    end
  end

  raise NoSuchValueError
end

# Returns the index of the value if the value exists and raises an error if the value dose not exist in the given array.
def recursive_binary_search(array, target, low, high)
  mid = (low + high) / 2
  # Raise error if the value of high is bigger that the one of low.
  raise NoSuchValueError if high < low
    # If the value in the middle is bigger that the target,
    # pass the value of `low` to the `low` parameter again and pass `mid - 1` to the `high` parameter.
    return recursive_binary_search(array, target, low, mid - 1) if (array[mid] > target)

    # If the value in the middle is smaller that the target,
    # pass the value of `high` to the `high` parameter again and pass `mid + 1` to the `low` parameter.
  return recursive_binary_search(array, target, mid + 1, high) if (array[mid] < target)

  return mid if (array[mid] == target)
end


# Returns true if the value exists and raises an error if the value dose not exist in the given array.
def recursive_binary_search2(array, target)
  low = 0
  high = array.length - 1
  raise NoSuchValueError if array.length == 0

  middle = (low + high) / 2
  # Check if the value in the middle is equal to the target.
  return middle if target == array[middle]

   # Check if the target is smaller or bigger than the value in the middle.
  if array[middle] < target
    # If the value in the middle is smaller that the target,
    # only pass values that are bigger than the value in the middle as array to the recursive_binary_search method. 
    recursive_binary_search(array[middle + 1, high], target)
  else
    # If the value in the middle is bigger that the target,
    # only pass values that are smaller than the value in the middle as array to the recursive_binary_search method. 
    srecursive_binary_search(array[low, middle - 1], target)
  end
end
